pretix_eth/exporter.py

```
@@ -0,0 +1,116 @@
+                              OrderRefund.REFUND_STATE_TRANSIT],
+                     widget=forms.CheckboxSelectMultiple,
+                     required=False
+                 )),
+            ]
+        )
+
+    def iterate_list(self, form_data):
+        payments = OrderPayment.objects.filter(
+            order__event__in=self.events,
+            state__in=form_data.get('payment_states', []),
+            provider='ethereum'
+        ).order_by('created')
+
+        refunds = OrderRefund.objects.filter(
+            order__event__in=self.events,
+            state__in=form_data.get('refund_states', []),
+            provider='ethereum'
+        ).order_by('created')
+
+        objs = sorted(list(payments) + list(refunds), key=lambda obj: obj.created)
+
+        headers = [
+            'Type', 'Event slug', 'Order', 'Payment ID', 'Creation date',
+            'Completion date', 'Status', 'Amount', 'Token', 'Wallet address'
+        ]
+
+        yield headers
+
+        yield self.ProgressSetTotal(total=len(objs))
+        for obj in objs:
+            time_zone = pytz.timezone(obj.order.event.settings.timezone)
+            if isinstance(obj, OrderPayment) and obj.payment_date:
+                completion_date = obj.payment_date.astimezone(
+                    time_zone).date().strftime('%Y-%m-%d')
+            elif isinstance(obj, OrderRefund) and obj.execution_date:
+                completion_date = obj.execution_date.astimezone(
+                    time_zone).date().strftime('%Y-%m-%d')
+            else:
+                completion_date = ''
+
+            if isinstance(obj, OrderPayment):
+                token = obj.info_data['currency_type']
+            elif isinstance(obj, OrderRefund):
+                token = obj.payment.info_data['currency_type']
+            else:
+                token = ''
+
+            if isinstance(obj, OrderPayment) and WalletAddress.objects.filter(
```

An alternate way to do this which *might* be cleaner and easier to reason about:

```python
def payment_to_row(payment: OrderPayment) -> ...:
    ...

def refund_to_row(refund: OrderRefund) -> ...:
    ...


for obj in objs:
    if isinstance(OrderPayment):
        row = payment_to_row(obj)
    elif isintance(OrderRefund):
        row = refund_to_row(obj)
    else:
        raise TypeError(...)
```

---

Good point, makes code a lot clearer. However, I think it is better to raise an invariant exception since such behavior not really possible.